File: examples/nodejs/consumer.js
#!/usr/bin/env node
/**
 * Kafka Consumer Example (Node.js)
 * 
 * This script consumes messages from a Kafka topic using KafkaJS.
 */

const { Kafka } = require('kafkajs');

// Configuration
const BOOTSTRAP_SERVERS = ['kafka-kafka-bootstrap.kafka.svc:9092']; // Internal access
// const BOOTSTRAP_SERVERS = ['<your-loadbalancer-endpoint>:9094']; // External access
const TOPIC = 'demo-topic';
const CLIENT_ID = 'nodejs-consumer';
const GROUP_ID = 'nodejs-consumer-group';

// Create Kafka client
const kafka = new Kafka({
  clientId: CLIENT_ID,
  brokers: BOOTSTRAP_SERVERS,
  retry: {
    retries: 5,
    initialRetryTime: 300,
  },
});

// Create consumer
const consumer = kafka.consumer({
  groupId: GROUP_ID,
  sessionTimeout: 30000,
  heartbeatInterval: 3000,
});

let messageCount = 0;

async function main() {
  try {
    // Connect consumer
    console.log('🔄 Connecting to Kafka...');
    await consumer.connect();
    console.log(`✓ Connected to Kafka at ${BOOTSTRAP_SERVERS.join(', ')}`);

    // Subscribe to topic
    await consumer.subscribe({
      topic: TOPIC,
      fromBeginning: true,
    });
    console.log(`✓ Subscribed to topic '${TOPIC}' with group '${GROUP_ID}'`);
    console.log('\n📥 Waiting for messages... (Press Ctrl+C to exit)\n');

    // Consume messages
    await consumer.run({
      eachMessage: async ({ topic, partition, message }) => {
        messageCount++;

        // Parse message
        const value = JSON.parse(message.value.toString());
        const headers = {};
        if (message.headers) {
          for (const [key, val] of Object.entries(message.headers)) {
            headers[key] = val.toString();
          }
        }

        // Display message
        console.log('='.repeat(60));
        console.log(`Message #${messageCount}`);
        console.log(`  Partition: ${partition}`);
        console.log(`  Offset: ${message.offset}`);
        console.log(`  Timestamp: ${message.timestamp}`);
        console.log(`  Key: ${message.key ? message.key.toString() : 'null'}`);
        console.log(`  Message ID: ${value.message_id || 'N/A'}`);
        console.log(`  Content: ${value.content || 'N/A'}`);
        console.log(`  Sent At: ${value.timestamp || 'N/A'}`);

        if (value.metadata) {
          console.log('  Metadata:');
          for (const [key, val] of Object.entries(value.metadata)) {
            console.log(`    ${key}: ${val}`);
          }
        }

        if (Object.keys(headers).length > 0) {
          console.log('  Headers:');
          for (const [key, val] of Object.entries(headers)) {
            console.log(`    ${key}: ${val}`);
          }
        }

        console.log('='.repeat(60) + '\n');
      },
    });

  } catch (error) {
    console.error(`\n✗ Error: ${error.message}`);
    process.exit(1);
  }
}

// Handle interruption
process.on('SIGINT', async () => {
  console.log('\n⚠ Interrupted by user');
  console.log(`\n📊 Total messages consumed: ${messageCount}`);
  console.log('\n🔄 Disconnecting consumer...');
  await consumer.disconnect();
  console.log('✓ Consumer disconnected');
  process.exit(0);
});

// Run main function
main().catch(console.error);

File: examples/python/consumer.py
#!/usr/bin/env python3
"""
Kafka Consumer Example

This script consumes messages from a Kafka topic.
"""

import json
from kafka import KafkaConsumer
from kafka.errors import KafkaError

# Configuration
BOOTSTRAP_SERVERS = ['kafka-kafka-bootstrap.kafka.svc:9092']  # Internal access
# BOOTSTRAP_SERVERS = ['<your-loadbalancer-endpoint>:9094']   # External access
TOPIC = 'demo-topic'
GROUP_ID = 'python-consumer-group'

def create_consumer():
    """Create and return a Kafka consumer."""
    try:
        consumer = KafkaConsumer(
            TOPIC,
            bootstrap_servers=BOOTSTRAP_SERVERS,
            group_id=GROUP_ID,
            # Deserialize messages
            value_deserializer=lambda m: json.loads(m.decode('utf-8')),
            # Start from earliest message if no offset exists
            auto_offset_reset='earliest',
            # Commit offsets automatically
            enable_auto_commit=True,
            auto_commit_interval_ms=1000,
            # Consumer session timeout
            session_timeout_ms=30000,
        )
        print(f"✓ Connected to Kafka at {BOOTSTRAP_SERVERS}")
        print(f"✓ Subscribed to topic '{TOPIC}' with group '{GROUP_ID}'")
        print("\n📥 Waiting for messages... (Press Ctrl+C to exit)\n")
        return consumer
    except KafkaError as e:
        print(f"✗ Failed to create consumer: {e}")
        raise

def consume_messages(consumer):
    """Consume and process messages from Kafka."""
    message_count = 0
    
    try:
        for message in consumer:
            message_count += 1
            
            # Extract message details
            partition = message.partition
            offset = message.offset
            timestamp = message.timestamp
            value = message.value
            
            # Display message
            print(f"{'='*60}")
            print(f"Message #{message_count}")
            print(f"  Partition: {partition}")
            print(f"  Offset: {offset}")
            print(f"  Timestamp: {timestamp}")
            print(f"  Message ID: {value.get('message_id', 'N/A')}")
            print(f"  Content: {value.get('content', 'N/A')}")
            print(f"  Sent At: {value.get('timestamp', 'N/A')}")
            
            if 'metadata' in value:
                print(f"  Metadata:")
                for key, val in value['metadata'].items():
                    print(f"    {key}: {val}")
            
            print(f"{'='*60}\n")
            
    except KeyboardInterrupt:
        print("\n⚠ Interrupted by user")
    finally:
        print(f"\n📊 Total messages consumed: {message_count}")

def main():
    """Main function to consume messages."""
    consumer = create_consumer()
    
    try:
        consume_messages(consumer)
    finally:
        # Close consumer
        print("\n🔄 Closing consumer...")
        consumer.close()
        print("✓ Consumer closed")

if __name__ == '__main__':
    main()
